docs/usage/deriving-data-selectors.md

```diff
@@ -0,0 +1,638 @@
+
+```js
+import { makeSelectItemsByCategory } from './categoriesSlice'
+
+const makeMapState = (state, ownProps) => {
+  // closure - create a new unique selector instance here,
+  // and this will run once for every component instance
+  const selectItemsByCategory = makeSelectItemsByCategory()
+
+  const realMapState = (state, ownProps) => {
+    return {
+      itemsByCategory: selectItemsByCategory(state, ownProps.category)
+    }
+  }
+}
+
+export default connect(makeMapState)(CategoryList)
+```
+
+## Using Selectors Effectively
+
+While selectors are a common pattern in Redux applications, they are often misused or misunderstood. Here are some guidelines for using selector functions correctly.
+
+### Define Selectors Alongside Reducers
+
+Selector functions are often defined in the UI layer, directly inside of `useSelector` calls. However, this means that there can be repetition between selectors defined in different files, and the functions are anonymous.
+
+Like any other function, you can extract an anonymous function outside the component to give it a name:
+
+```js
+// highlight-next-line
+const selectTodos = state => state.todos
+
+function TodoList() {
+  // highlight-next-line
+  const todos = useSelector(selectTodos)
+}
+```
+
+However, multiple parts of the application may want to use the same lookups. Also, conceptually, we may want to keep the knowledge of how the `todos` state is organized as an implementation detail inside the `todosSlice` file, so that it's all in one place.
+
+Because of this, **it's a good idea to define reusable selectors alongside their corresponding reducers**. In this case, we could export `selectTodos` from the `todosSlice` file:
+
+```js title="src/features/todos/todosSlice.js"
+import { createSlice } from '@reduxjs/toolkit'
+
+const todosSlice = createSlice({
+  name: 'todos',
+  initialState: [],
+  reducers: {
+    todoAdded(state, action) {
+      state.push(action.payload)
+    }
+  }
+})
+
+export const { todoAdded } = todosSlice.actions
+export default todosSlice.reducer
+
+// highlight-start
+// Export a reusable selector here
+export const selectTodos = state => state.todos
+// highlight-end
+```
+
+That way, if we happen to make an update the structure of the todos slice state, the relevant selectors are right here and can be updated at the same time, with minimal changes to any other parts of the app.
+
+### Balance Selector Usage
+
+It's possible to add _too many_ selectors to an application. **Adding a separate selector function for every single field is not a good idea!** That ends up turning Redux into something resembling a a Java class with getter/setter functions for every field. It's not going to _improve_ the code, and it's probably going to make the code _worse_ - maintaining all those extra selectors is a lot of additional effort, and it will be harder to trace what values are being used where.
```

Comment:
```suggestion
It's possible to add _too many_ selectors to an application. **Adding a separate selector function for every single field is not a good idea!** That ends up turning Redux into something resembling a Java class with getter/setter functions for every field. It's not going to _improve_ the code, and it's probably going to make the code _worse_ - maintaining all those extra selectors is a lot of additional effort, and it will be harder to trace what values are being used where.
```